IAP brand & client

drebes opened the pull request:
<!-- AUTOCHANGELOG for Downstream PRs.

EXTERNAL CONTRIBUTORS: Ignore please - your reviewer will handle.

INTERNAL CONTRIBUTORS AND REVIEWERS: See .ci/RELEASE_NOTES_GUIDE.md
for writing good release notes.

NO CHANGELOG NOTE: Please add "changelog: no-release-note" label to this PR.

Otherwise, fill the template out (replace the heading).
You can add more release notes if you want more than one CHANGELOG entry for
this PR, but make sure not to indent notes and to leave newlines between
code blocks for Markdown's sake.

For Terraform PRs, we use the following "release-note:" headings
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
-->
Fixes terraform-providers/terraform-provider-google#1287

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_iap_brand`
```
```release-note:new-resource
`google_iap_client`
```

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3135-old..auto-pr-3135) ( 9 files changed, 923 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3135-old..auto-pr-3135) ( 9 files changed, 923 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3135-old..auto-pr-3135) ( 2 files changed, 118 insertions(+))

drebes commented on templates/terraform/examples/iap_identity_aware_proxy_client.tf.erb:
```diff
@@ -0,0 +1,21 @@
+}
+
+resource "google_iap_brand" "project_brand" {
+  support_email     = "support@example.com"
```
This acceptance test won't succeed unless the specified group is owned by the service account running the test.

drebes commented on third_party/terraform/tests/resource_iap_brand_test.go:
```diff
@@ -0,0 +1,54 @@
+}
+
+resource "google_iap_brand" "project_brand" {
+  support_email     = "support@example.com"
```
This acceptance test won't succeed unless the specified group is owned by the service account running the test.

drebes replied:
Service accounts cannot be specified. Users can be created but not with the Google provider, only the 3rd-party G Suite one.

drebes replied:
I've added more details to the supportEmail description to make it clear what are the valid options.

drebes commented:
Fixes terraform-providers/terraform-provider-google#1287

danawillow commented:
drive-by: the "fixes" statement won't automatically close the issue if it's in a comment- it needs to be in the PR description or a commit (I usually do the PR description like in https://github.com/GoogleCloudPlatform/magic-modules/pull/3063)

slevenick commented on products/iap/api.yaml:
```diff
@@ -135,3 +135,66 @@ objects:
+      - !ruby/object:Api::Type::String
+        name: 'name'
+        description: |
+          Output only. Identifier of the brand. NOTE: GCP project number achieves
```
Can we document the format of `name`? In the generated documentation we have the import format listed as just `{name}` which will be difficult to figure out if we don't list the actual format of `name` somewhere

drebes replied:
Done.

slevenick commented on products/iap/api.yaml:
```diff
@@ -135,3 +135,66 @@ objects:
+      - !ruby/object:Api::Type::String
+        name: 'brand'
+        description: |
+          Identifier of the brand to which this `IdentityAwareProxyClients``
```
Can we document the format of `brand` here? We show the import id of this resource as `{brand}/identityAwareProxyClients/{client_id}` but a user may not know the format of `brand`.

drebes replied:
Done.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3135-old..auto-pr-3135) ( 9 files changed, 938 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3135-old..auto-pr-3135) ( 9 files changed, 938 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3135-old..auto-pr-3135) ( 2 files changed, 118 insertions(+))

pvillard31 commented:
_Sorry for the comment/question, feel free to redirect me to the right place if I should ask this somewhere else._

If I correctly understand this PR, its purpose is to provide a way to create the "APIs & Services / OAuth consent screen". However, I do not see how to specify information such as authorised domains, homepage link, etc. Am I missing something?

Also I see that client ID and client secret are made available but how to specify  "Authorised JavaScript origins" and "Authorised redirect URIs". Is it out of scope of this PR?

drebes commented on products/iap/api.yaml:
```diff
@@ -135,3 +135,66 @@ objects:
+          Whether the brand is only intended for usage inside the GSuite organization only.
+        output: true
+  - !ruby/object:Api::Resource
+    name: 'IdentityAwareProxyClient'
```
Indeed looks much better. Changed.

drebes commented on products/iap/api.yaml:
```diff
@@ -135,3 +135,66 @@ objects:
+    self_link: '{{name}}'
+    input: true
+    description: |
+      OAuth brand data.
```
Done.

drebes commented on templates/terraform/extra_schema_entry/iap_identity_aware_proxy_client.erb:
```diff
@@ -0,0 +1,4 @@
+"client_id": {
```
Done. post_create is still needed as the name/clientId is not returned on the create call, though.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3135-old..auto-pr-3135) ( 9 files changed, 939 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3135-old..auto-pr-3135) ( 9 files changed, 939 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3135-old..auto-pr-3135) ( 2 files changed, 118 insertions(+))

slevenick commented:
https://github.com/GoogleCloudPlatform/magic-modules/pull/3162 Should fix the issues I ran into around the beta provider and base path names.

I'm happy with the current implementation, I would just like to get integration tests working for this so that we can run them via CI.

I think all we need is to add a new environment variable to `provider_test.go` that will represent the organization domain. Then we can depend on `support@{org-domain}` existing and the service account running the tests to be an admin for that group email (I'll go set this up manually)

slevenick commented:
Hey @drebes do you think you will have a chance to add support for those tests via the environment variable sometime soon?

I can also pick this up and add the changes to your branch, just let me know if you want me to do that

drebes commented:
Sure I can add them (sorry, I didn't understand "we need is to add a new environment variable" to mean me, but I can look into it.)

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3135-old..auto-pr-3135) ( 10 files changed, 950 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3135-old..auto-pr-3135) ( 10 files changed, 950 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3135-old..auto-pr-3135) ( 2 files changed, 118 insertions(+))

drebes commented:
Added the configurable domain for the acceptance tests.

You need to create a group "support@<acceptance_test_domain>" and make the acceptance test service account an owner of such group, and then set `GOOGLE_ORG_DOMAIN` as <acceptance_test_domain>.

slevenick commented:
Awesome, looks good! I got it to pass locally, so I'll modify the CI test org to add this. Can you rebase this from master, looks like a couple things are out of sync

drebes commented:
There you go. Rebased with the collapsed sidebar changes that were causing the conflict. Which also made me realized that I hadn't updated the sidebar for `iap_client` since I had updated it to the short form. Should be good now.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3135-old..auto-pr-3135) ( 10 files changed, 950 insertions(+), 2 deletions(-))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3135-old..auto-pr-3135) ( 10 files changed, 950 insertions(+), 2 deletions(-))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3135-old..auto-pr-3135) ( 2 files changed, 118 insertions(+))
